### PR TITLE
🎉 New Source: Asana

### LINKED ISSUE #1969: Integration with Asana
## Tell us about the new integration you’d like to have
Which source and which destination? Which frequency?

## Describe the context around this new integration
Which team in your company wants this integration, what for? This helps us understand the use case.

## Describe the alternative you are considering or using
What are you considering doing if you don’t have this integration through Airbyte?

Would be great to add Asana to your sources.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368211269044) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by Zirochkaa**
## What
Closes #1969

## Pre-merge Checklist
* [x] _Run integration tests_
* [x] _Publish Docker images_

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368060255977) by [Unito](https://www.unito.io)

### CODE REVIEW

File: airbyte-integrations/connectors/source-asana/integration_tests/abnormal_state.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (keu) [inline airbyte-integrations/connectors/source-asana/integration_tests/abnormal_state.json:1]: if you don't use incremental tests then you can drop this file

File: airbyte-integrations/connectors/source-asana/integration_tests/catalog.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
+{
```
Comment (keu) [inline airbyte-integrations/connectors/source-asana/integration_tests/catalog.json:1]: you can drop this file

File: airbyte-integrations/connectors/source-asana/integration_tests/sample_config.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (keu) [inline airbyte-integrations/connectors/source-asana/integration_tests/sample_config.json:1]: this should be a sample config

File: airbyte-integrations/connectors/source-asana/integration_tests/sample_state.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (keu) [inline airbyte-integrations/connectors/source-asana/integration_tests/sample_state.json:1]: this should be a sample state, real stream name, column and some value

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    """
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        workspaces_stream = Workspaces(authenticator=self.authenticator)
```
Comment (keu) [inline airbyte-integrations/connectors/source-asana/source_asana/streams.py:105]: why can't we use sliced stream here? just override `request_params` and `stream_slice` methods

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    def path(self, **kwargs) -> str:
+        return "projects"
+
+
+class Sections(AsanaStream):
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        project_gid = stream_slice["project_gid"]
+        return f"projects/{project_gid}/sections"
+
+    def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
```
Comment (keu) [inline airbyte-integrations/connectors/source-asana/source_asana/streams.py:155]: better use stream_slices and remove read_records override

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        for project in projects_stream.read_records(sync_mode=SyncMode.full_refresh):
+            yield from super().read_records(stream_slice={"project_gid": project["gid"]}, **kwargs)
+
+
+class Stories(AsanaStream):
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        task_gid = stream_slice["task_gid"]
+        return f"tasks/{task_gid}/stories"
+
+    def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
```
Comment (keu) [inline airbyte-integrations/connectors/source-asana/source_asana/streams.py:166]: same

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+class Tags(AsanaStreamWorkspacePagination):
+    def path(self, **kwargs) -> str:
+        return "tags"
+
+
+class Tasks(AsanaStream):
+    def path(self, **kwargs) -> str:
+        return "tasks"
+
+    def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
```
Comment (keu) [inline airbyte-integrations/connectors/source-asana/source_asana/streams.py:181]: same

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+        return f"{prop}.gid"
+
+
+class Teams(AsanaStream):
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        workspace_gid = stream_slice["workspace_gid"]
+        return f"organizations/{workspace_gid}/teams"
+
+    def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
```
Comment (keu) [inline airbyte-integrations/connectors/source-asana/source_asana/streams.py:211]: same

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+        for workspace in workspaces_stream.read_records(sync_mode=SyncMode.full_refresh):
+            yield from super().read_records(stream_slice={"workspace_gid": workspace["gid"]}, **kwargs)
+
+
+class TeamMemberships(AsanaStream):
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        team_gid = stream_slice["team_gid"]
+        return f"teams/{team_gid}/team_memberships"
+
+    def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
```
Comment (keu) [inline airbyte-integrations/connectors/source-asana/source_asana/streams.py:222]: same

File: docs/integrations/connector-health.md [hunk 11]
```diff
@@ -37,6 +37,7 @@ Clicking on a status will take you to a summary of past tests that link to the G
 | HTTP Request | [![source-http-request](https://img.shields.io/endpoint?url=https%3A%2F%2Fstatus-api.airbyte.io%2Ftests%2Fsummary%2Fsource-http-request%2Fbadge.json)](https://status-api.airbyte.io/tests/summary/source-http-request) |  |  |
 | Hubspot | [![source-hubspot-singer](https://img.shields.io/endpoint?url=https%3A%2F%2Fstatus-api.airbyte.io%2Ftests%2Fsummary%2Fsource-hubspot%2Fbadge.json)](https://status-api.airbyte.io/tests/summary/source-hubspot) | ✅ |  |
 | Klaviyo | [![source-klaviyo](https://img.shields.io/endpoint?url=https%3A%2F%2Fstatus-api.airbyte.io%2Ftests%2Fsummary%2Fsource-klaviyo%2Fbadge.json)](https://status-api.airbyte.io/tests/summary/source-klaviyo) | | |
+| Klaviyo | [![source-asana](https://img.shields.io/endpoint?url=https%3A%2F%2Fstatus-api.airbyte.io%2Ftests%2Fsummary%2Fsource-asana%2Fbadge.json)](https://status-api.airbyte.io/tests/summary/source-asana) | | |
```
Comment (keu) [inline docs/integrations/connector-health.md:40]: what a nice disguise  :)

**Review by keu — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-asana/integration_tests/sample_state.json (hunk 4 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/integration_tests/sample_state.json:1]: Asana source doesn't support incremental sync for streams in this PR. So should I just delete this file?

File: docs/integrations/connector-health.md (hunk 11 above)
Comment (Zirochkaa) [reply docs/integrations/connector-health.md:40]: Oops, I thought I changed every one of them 😅

File: airbyte-integrations/connectors/source-asana/integration_tests/abnormal_state.json (hunk 1 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/integration_tests/abnormal_state.json:1]: Done.

File: airbyte-integrations/connectors/source-asana/integration_tests/catalog.json (hunk 2 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/integration_tests/catalog.json:1]: Done.

File: docs/integrations/connector-health.md (hunk 11 above)
Comment (Zirochkaa) [reply docs/integrations/connector-health.md:40]: Done.

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py (hunk 5 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py:105]: Done.

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+        # Always return an empty generator just in case no records were ever yielded
+        yield from []
+
+
+class CustomFields(AsanaStream):
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        workspace_gid = stream_slice["workspace_gid"]
+        return f"workspaces/{workspace_gid}/custom_fields"
+
+    def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
```
Comment (Zirochkaa) [inline airbyte-integrations/connectors/source-asana/source_asana/streams.py:139]: better use stream_slices and remove read_records override

Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py:139]: Done.

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py (hunk 9 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py:211]: Done.

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py (hunk 8 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py:181]: Done.

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py (hunk 10 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py:222]: Done.

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py (hunk 7 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py:166]: Done.

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py (hunk 6 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py:155]: Done.

File: airbyte-integrations/connectors/source-asana/integration_tests/sample_state.json (hunk 4 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/integration_tests/sample_state.json:1]: I just delete this file.

File: airbyte-integrations/connectors/source-asana/integration_tests/sample_config.json (hunk 3 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/integration_tests/sample_config.json:1]: What does this mean? Like we can't put actual `access_token` there. For now I'll change content of this file to this:
```
{
  "access_token": "TODO"
}
```

File: airbyte-integrations/connectors/source-asana/CHANGELOG.md [hunk 13]
```diff
@@ -0,0 +1,4 @@
+# Changelog
+
+## 0.1.1
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-asana/CHANGELOG.md:3]:
```suggestion
## 0.1.0
```

**Review by sherifnada — APPROVED** [review]

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, any] = None,
+        next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+
+        # Asana pagination could be from 1 to 100.
+        params = {"limit": 100}
```
Comment (keu) [inline airbyte-integrations/connectors/source-asana/source_asana/streams.py:58]: let's move `100` to `AsanaStream.page_size`

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+    """
+
+    def stream_slices(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        projects_stream = Projects(authenticator=self.authenticator)
+        projects_slices = projects_stream.stream_slices(sync_mode=SyncMode.full_refresh)
```
Comment (keu) [inline airbyte-integrations/connectors/source-asana/source_asana/streams.py:136]: again, why do we need slices here?

**Review by keu — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py (hunk 14 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py:58]: Done.

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+
+    def stream_slices(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        workspaces_stream = Workspaces(authenticator=self.authenticator)
+        workspaces_slices = workspaces_stream.stream_slices(sync_mode=SyncMode.full_refresh)
+        for workspace_slice in workspaces_slices:
```
[the remark this answers is not in the document]
Comment (sherifnada) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py:118]: this is a good question - missed this in review. Good catch @keu

[the remark this answers is not in the document]
Comment (keu) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py:118]: @sherifnada actually after discussion with @Zirochkaa we came to conclusion that we need to have method in CDK that will just read (call stream slices and read records internally)

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+
+        params.update(self.get_opt_fields())
+
+        if next_page_token:
+            params.update(next_page_token)
+
+        return params
+
+    def get_opt_fields(self) -> MutableMapping[str, str]:
+        opt_fields = list()
```
[the remark this answers is not in the document]
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py:68]: Done.

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py:118]: Done.

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py (hunk 15 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py:136]: Done.

**Review by keu — APPROVED** [review]
lgtm

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py [hunk 18]
```diff
@@ -177,14 +191,10 @@ def stream_slices(
         cursor_field: List[str] = None,
         stream_state: Mapping[str, Any] = None
     ) -> Iterable[Optional[Mapping[str, Any]]]:
-        tasks_stream = Tasks(authenticator=self.authenticator)
-        tasks_slices = tasks_stream.stream_slices(sync_mode=SyncMode.full_refresh)
-        for task_slice in tasks_slices:
-            for task in tasks_stream.read_records(sync_mode=SyncMode.full_refresh, stream_slice=task_slice):
-                yield {"task_gid": task["gid"]}
+        yield from self.read_stream(stream_class=Tasks, slice_field="task_gid")
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-asana/source_asana/streams.py:194]: the `slice_field` param is a little off here -- this method name implies it should just read the stream but the parameter contains some functionality related to slices. Could we rename this method to be something like `read_slices_from_records` or something?

**Review by sherifnada — APPROVED** [review]

File: airbyte-integrations/connectors/source-asana/source_asana/streams.py (hunk 18 above)
Comment (Zirochkaa) [reply airbyte-integrations/connectors/source-asana/source_asana/streams.py]: Done.

**Review by sherifnada — APPROVED** [review]

**Comment by marcosmarxm** [comment]
@sherifnada looks we need Asana icon without is breaking master.

**Comment by Zirochkaa** [comment]
@marcosmarxm according to [this](https://github.com/airbytehq/airbyte/issues/3588#issuecomment-848103928) and [this](https://github.com/airbytehq/airbyte/issues/3588#issuecomment-848113340) comments everything is ok right now, right?

**Comment by marcosmarxm** [comment]
@Zirochkaa you are correct

**Pull request merged**